File: java-webflux/src/main/java/com/ederfmatos/javawebflux/usecase/CreateTransaction.java
package com.ederfmatos.javawebflux.usecase;

import com.ederfmatos.javawebflux.entity.Transaction;
import com.ederfmatos.javawebflux.enums.Currency;
import com.ederfmatos.javawebflux.gateway.CurrencyGateway;
import com.ederfmatos.javawebflux.repository.TransactionRepository;
import java.math.BigDecimal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
@RequiredArgsConstructor
public class CreateTransaction {
    private final CurrencyGateway currencyGateway;
    private final TransactionRepository transactionRepository;

    public Mono<Output> execute(Input input) {
        return currencyGateway.getCurrencies()
                .map(currencies -> {
                    BigDecimal currency = currencies.get(input.currency);
                    var totalAmount = currency.multiply(input.amount);
                    return Transaction.create(input.description, input.amount, input.currency, totalAmount);
                })
                .flatMap(transaction -> transactionRepository.create(transaction).thenReturn(transaction))
                .map(transaction -> new Output(transaction.id()));
    }

    public record Input(String description, BigDecimal amount, Currency currency) {
    }

    public record Output(String id) {
    }
}

File: kotlin-webflux/src/main/kotlin/com/ederfmatos/kotlinwebflux/usecase/CreateTransaction.kt
package com.ederfmatos.kotlinwebflux.usecase

import com.ederfmatos.kotlinwebflux.entity.Transaction
import com.ederfmatos.kotlinwebflux.enums.Currency
import com.ederfmatos.kotlinwebflux.gateway.CurrencyGateway
import com.ederfmatos.kotlinwebflux.repository.TransactionRepository
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import java.math.BigDecimal

@Component
class CreateTransaction(
    private val currencyGateway: CurrencyGateway,
    private val transactionRepository: TransactionRepository,
) {
    fun execute(input: Input): Mono<Output> {
        return currencyGateway.getCurrencies()
            .map { currencies ->
                val amount = currencies[input.currency]!!.multiply(input.amount)
                Transaction.create(
                    description = input.description,
                    totalAmount = amount,
                    amount = input.amount,
                    currency = input.currency,
                )
            }.flatMap { transaction -> transactionRepository.create(transaction).thenReturn(transaction) }
            .map { transaction -> Output(transaction.id) }
    }

    data class Input(val description: String, val amount: BigDecimal, val currency: Currency)
    data class Output(val id: String)
}